src/gz/read.rs

```diff
@@ -90,13 +90,22 @@ impl<R: Read + Write> Write for GzEncoder<R> {
     }
 }
 
-/// A gzip streaming decoder
+/// A decoder for the first member of a [gzip file].
 ///
 /// This structure exposes a [`Read`] interface that will consume compressed
 /// data from the underlying reader and emit uncompressed data.
-/// Use [`MultiGzDecoder`] if your file has multiple streams.
 ///
-/// [`Read`]: https://doc.rust-lang.org/std/io/trait.Read.html
+/// After reading the first member of a gzip file (which is often, but not
+/// always, the only member), this reader will return Ok(0) even if there
+/// are more bytes available in the underlying reader. If you want to be sure
+/// not to drop bytes on the floor, call `into_inner()` after Ok(0) to
+/// recover the underlying reader.
```

Comment:
This doesn't apply to `read::GzDecoder`. The caller loses any bytes between the end of the gzip data and the end of the last read done by the `GzDecoder`. Not losing these bytes is the benefit of using `bufread::GzDecoder`.

```suggestion
/// After reading a single member of the gzip data this reader will return
/// Ok(0) even if there are more bytes available in the underlying reader.
/// `GzDecoder` may have read additional bytes past the end of the gzip data.
/// If you need the following bytes, wrap the `Reader` in a `std::io::BufReader` 
/// and use `bufread::GzDecoder` instead.
```

Reply:
Thanks for this clarification, this is subtle.
I must admit that I am still absolutely confused by the many different implementations of what seems to be types with the same name.

Reply:
Generally the `Read` version is the natural approach for sync code, and the `Write` version is the natural approach for async code.

The `Read` version has the above problem, which the `BufRead` version solves. The `Read` version then delegates to the `BufRead` version. The `Read` and `BufRead` versions could have been the same, but at the cost that a `Read: BufRead` type would get buffered twice, since there is no easy way to convert to a `BufRead` only if `!BufRead`.

What should probably have been better is for the `Read` version to not have the `as_ref`, `as_mut`, and `into_inner` methods to extract the underlying `Read`. Since they do exist, we should add a note that these methods do not work as expected. But that can be a separate PR.

Reply:
I have checked again and didn't see this very useful information in the top-level crate documentation, even though I think it definitely has a place there.
If you agree, would you mind sending a PR for this? If there isn't enough time, please let me know as well and I'd take it on me to incorporate this knowledge, stating you as co-author. Thanks a lot.